content/post/2020-02-07-processx-blocked-sigchld-fix/index.Rmd

```diff
@@ -0,0 +1,442 @@
+---
+slug: processx-blocked-sigchld
+title: "Debugging: Signals and Subprocesses"
+authors:
+  - Gábor Csárdi
+date: "2020-02-07"
+tags:
+- package development
+- debugging
+output: 
+  html_document:
+    keep_md: true
+---
+
+```{r include = FALSE}
+options(width = 60)
+```
+
+This is a short story about a non-trivial bug in the processx package,
+and how I fixed it. It is a good showcase of the some debugging tools.
+
+## The bug
+
+[processx](https://processx.r-lib.org) is an R package to start and manage
+external processes. It is used by the [callr](https://call.r-lib.org)
+package to run code in another R session. The
+[original bug](https://github.com/r-lib/processx/issues/240) report has a
+nice, clean [reproducible example](https://reprex.tidyverse.org/):
+
+> https://github.com/r-lib/processx/pull/237 seems to solve
+> https://github.com/r-lib/processx/issues/236 for `processx`,
+> but not for `callr`.
+> 
+> ```r
+> fun <- function() {
+>   parallel::mclapply(1:2, function(x) x)
+> }
+> env <- c(
+>   callr::rcmd_safe_env(),
+>   PROCESSX_NOTIFY_OLD_SIGCHLD = "true"
+> )
+> tmp <- callr::r(fun, env = env, show = TRUE)
+> #> Error while shutting down parallel: unable to terminate some child processes
+```
+
+## A red herring
+
+The references to [PR #237](https://github.com/r-lib/processx/pull/237) and
+[issue #236](https://github.com/r-lib/processx/issues/236) are a red
+herring, and a tricky one! The error message is exactly the same there,
```

Comment:
Add a link to https://en.m.wikipedia.org/wiki/Red_herring? I had to look up the definition 🙈

Reply:
I think we can just let people google. :)

Reply:
I think it can make non native speakers feel inadequate

Reply:
OK, I'll link then.